### ISSUE TITLE
kmax affecting non-linear P(k)

### ISSUE THREAD

**Issue by alexander-mead** [cmbant/CAMB#53]
I had another thought about kmax: The results from HMcode depend on kmax of the linear spectrum because HMcode uses things like sigma(R), sigmaV(R), neff(R), which are all technically integrals over all k of the linear P(k). HMcode extrapolates the power beyond kmax (via a power law) to do these integrals, but the slope of this power law will be quite different if kmax=1 or kmax=10 are used. When I have been using the code I always used kmax=100 with the idea that this would produce convergent results at k=10. Is it worth putting some sort of warning in the code to stop too low values of kmax being used? I have had a few emails about this recently and I worry that a desire for people to speed up MCMC chains is leading to people scrimping on kmax, which may be affecting their results.

 I note that the same behaviour is true of HALOFIT for the same reasons.

**Comment by cmbant** [comment]
Yes it might make sense to have a warning in e.g. set_matter_power at least. Is there an approximate guide, i.e. if I want it correct at k_dat, what k_max do I need? I'm just off for a couple of days now though.

From 10-100 is the linear spectrum not reasonably well fit by [1+b*ln(c*k)^2]P_R(0) (e..g eq 71 of https://cosmologist.info/teaching/EU/notes_structure.pdf or one of the old actual fitting formula), at least for LCDM? i.e. one could attempt to do a reasonable extrapolation without explicit calculation of the slow very high k modes (at the accuracy to which halofit works)?

**Comment by alexander-mead** [comment]
I was always using kmax=100 with the idea that this gave convergent results at k=10. I can't remember exactly what I did to check this though, or what I meant exactly by 'convergent'. Probably it had to do with numerical error relative to the errors in the emulators I was tuning HMcode to.

I had been using P(k) = P_max (ln(k)/ln(k_max))^2 (k/k_max)^(ns-4) for high-k extrapolation of the linear spectrum (not exactly the same as your formula), which looked reasonably good between k=10 and 100 if kmax=10 was set, but looks awful if kmax=1 was set, which is why I changed this to extrapolation via a power law. 

Just looking at my extrapolation formula now it looks a bit weird, since the log is being taken of a dimensionful quantity...

**Comment by cmbant** [comment]
For the moment 1.0.7 adds a warning to set_matter_power. Should look at better extrapolation for a future version.

**Comment by cmbant** [comment]
I had a quick look at the errors with kmax (relative to kmax=100), which look like a very odd function of redshift:
<img width="743" alt="kmaxerrors" src="https://user-images.githubusercontent.com/3958819/62349774-08dff300-b501-11e9-8be1-22f732284ee5.png">

Here I just did
```
kmax = np.logspace(-1,2,12)
print(kmax)
res=[]
for k in kmax:
    pars = camb.CAMBparams()
    pars.set_cosmology(H0=67.5, ombh2=0.022, omch2=0.122)
    pars.set_matter_power(np.arange(0,20,0.2),kmax=k, nonlinear=True)
    pars.InitPower.set_params(ns=0.965, As=2e-9)

    results = camb.get_results(pars)
    res += [results.get_matter_power_interpolator()]

_, axs = plt.subplots(2,2, figsize=(12,12))
for z,ax in zip([0,1,2,5], axs.reshape(-1)):
    for k,P in zip(kmax,res):
        kh=np.logspace(-3, np.log10(k))
        ax.semilogx(kh,P.P(z,kh)/res[-1].P(z,kh)-1)
    leg = ["%.2f"%f for f in kmax]
    ax.legend(leg)
    ax.set_xlabel('$k$')
    ax.set_ylabel(r'$\Delta P(k)/P(k)$')
    ax.set_title(r'$z=%s$'%z)`
```

Any idea what's going on here?

**Comment by cmbant** [comment]
Note, my understanding of how the code currently works is that HMCode internally uses kmax=100 (assuming rebin_pk=T as it is by default) even if CAMB's kmax is lower. This means that it's using CAMB's log-linear extrapolation in the linear spectrum beyond CAMB's kmax.

**Comment by alexander-mead** [comment]
Sorry for the long delay on this. I just remade your plots so that I could fully understand what is going on here: I changed the colour scheme for the lines and forced the y axes to all show the same range and converted to a % difference on y. I made two sets of results, one for HMcode (halofit_version='mead') and one for HALOFIT (halofit_version='takahashi')

Mead:
![image](https://user-images.githubusercontent.com/9140961/73771906-866e0f80-477f-11ea-8854-1d27a81ab76d.png)

Takahashi:
![image](https://user-images.githubusercontent.com/9140961/73772062-c634f700-477f-11ea-80ef-1f65049c1814.png)

Both sets of results look quite similar, which I think I would expect.  We also see that the error is worse for a fixed kmax as you go to high redshift. This I would expect since the non-linear power at higher redshift is governed by smaller and smaller scales the further back you go in redshift. These plots tell me that if I want numerical convergence in the non-linear power at z = 0 at k = 1 then I am okay with setting the CAMB kmax = 1, but if I want convergence at the same k=1 at higher z>0 then I should use a higher kmax.

What exactly were you concerned about here?

**Comment by cmbant** [comment]
Thanks for checking, it's probably fine. The way the error at fixed k increases to high redshift is a bit unintuitive given that the spectrum at fixed k is getting more linear so the overall correction should be smaller - though of course I agree that at high z the non-linear correction is all coming from smaller k. I see it does actually get smaller again above z~5.

![image](https://user-images.githubusercontent.com/3958819/73843392-7a815c80-4816-11ea-9d4f-b6f51ea27740.png)

For reference this is the fractional error in the non-linear correction:
![image](https://user-images.githubusercontent.com/3958819/73852672-8295c800-4827-11ea-82f2-9a2403d55a34.png)

**Comment by alexander-mead** [comment]
Okay. I think this all makes sense to me then. I don't think there are any problems here. I'm happy for this to be closed if you are. I do agree that this is a bit unintuitive. 

Also, when I fitted the model, it was only every done for z<2, and I wouldn't want to speculate too much on how accurate the results would be for the (smaller) non-linear correction at higher z. In general, it always seemed that getting a good model fit at higher z was more difficult, which I found surprising. In the first HMcode paper we speculated that this might be because the linear power spectrum effective index at the non-linear point changes a bit between z=0 and z=2 for standard cosmological models, and the linear power spectrum (in Delta^2(k) units) around the non-linear scale becomes flatter at higher z. A consequence of this would be that many scales are collapsing simultaneously, with the corollary that the structure-formation process would probably be more disorderly in the past compared to today and therefore that assumptions about virialised structures may be less valid.
